### DISCUSSION TITLE
How to Add Keyboard Events like "Delete" for selected object

### DISCUSSION THREAD

**Discussion by assentorp** [fabricjs/fabric.js#6990] [Q&A]
Hi, I'm using Reactjs and have implemented fabricjs. So far so good. Now I'm trying to add a simple keyboard event that deletes a selected object when you hit "Backspace" on the keyboard. If I'm editing text it shouldn't delete the textbox when I hit Backspace. How can I achieve this with Reactjs?

_I'm a designer who knows a bit of code, so I'm not a great developer... please be kind :-)_

**Comment by asturur** [comment]
React.js is not adding any difference here. Chiose your event library or no event library for keyboard events, then attach the event handler to the right spot ( document? window?).
The event handler could look like something like this:

```
if (rightKeyComboDetected) {
const activeObjects = fabricCanvas.getActiveObjects();
if (activeObjects[0].isEditing) return;
fabricCanvas.remove(...activeObjects);
}
```

**Reply by assentorp** [reply]
Thanks for the quick reply... again im not the best at this stuff, is this the right path?

```
  const deleteShortcut = (canvas) => {
    if (canvas.keyCode === 8) {
      const activeObjects = canvas.getActiveObjects();
      if (activeObjects[0].isEditing) return;
      canvas.remove(...activeObjects);
    }
  };
```

my HTML:
```
          <canvas id="canvas" onKeyDown={deleteShortcut}>
            Your browser doesn't support this. You should consider upgrading.
          </canvas>
```

**Reply by asturur** [reply]
i can give you a more detailed answer, but i need to sit with a  proper keyboard at the computer.

**Reply by asturur** [reply]
The canvas does not support a tabIndex i suppose and i think won't receive keyboard events in general.
You are using reactJS, so let's assume you have a canvas component:

```jsx

import React, { useRef, useEffect } from 'react';

const Canvas = () => {

const fabricCanvas = useRef(new fabric.Canvas())
const canvasEl = useRef();

// initialize your canvas
useEffect(() => {
  fabricCanvas.current.initialzie(canvasEl.current, { ...yourOptions });
}, [canvasEl.current]);

// add the key listener to window
useEffect(() => {
  window.addEventListener('keyup', (e) => {
    if (e.keycode === 'youknowwhat') {
      const canvas= fabricCanvas.current;
      const activeObjects = canvas.getActiveObjects();
      if (activeObjects[0].isEditing) return;
      canvas.remove(...activeObjects);
    }    
  });
}, [])

return <canvas ref={canvasEl } ></canvas>;
}
```

Something like this.

Then how do you make your fabricCanvas instance available to a whole app without using spaghetti code, that is another story, i would look into the useContext hook and the context api in general for react, but that you can look into when this snippet works and is clear. The you can move to the next step.

**Reply by assentorp** [reply]
Hi again asturur, thanks for your reply and trying to help me, really appreciate it. I can't figure out how to put that into my spaghetti code.  I'm already using UseEffect to initialize my canvas and now I'm here (this is just a snippet):

```
export default () => {
  const [canvas, setCanvas] = useState("");
  const canvasWidth = 1200;
  const canvasHeight = 1200;

  useEffect(() => {
    setCanvas(
      new fabric.Canvas("canvas", {
        width: canvasWidth,
        height: canvasHeight,
        backgroundColor: "white",
        preserveObjectStacking: true,
      })
    );
    fabric.Object.prototype.transparentCorners = false;
    fabric.Object.prototype.cornerColor = "#ff6c14";
    fabric.Object.prototype.cornerStyle = "rect";
    fabric.Object.prototype.cornerSize = "8";
    fabric.Object.prototype.borderColor = "#ff6c14";

  window.addEventListener('keyup', (e) => {
    if (e.keycode === 8) {
      const canvas = canvas.current;
      const activeObjects = canvas.getActiveObjects();
      if (activeObjects[0].isEditing) return;
      canvas.remove(...activeObjects);
    }    
  });
  }, []);

  return (
    <div>
          <canvas id="canvas">
            Your browser doesn't support this. You should consider upgrading.
          </canvas>
</div>
);
};

```
Why won't it accept my current canvas here?

**Reply by asturur** [reply]
`.current` is from use useRef from react standard hooks. I put my fabricCanvas instance in a ref, you put it in state.

So in your case removing `.current` should make it, but also, since everything is used inside useEffect, you do not really need to have canvas defined outside of it.

so the whole setCanvas seems to much, also because setCanvas is a setState, is going to re-render your component, but also the canvas instance isn't influencing the react managed dom, so you do not really need to trigger a re-render.

I do not know which is your current experience with react lifecycles, and things get hairy quickly.

**Reply by assentorp** [reply]
I decided to refactor some of the code in order to get it to work. The shortcuts work flawlessly now (THANKS SO MUCH!!)... however it broke my resize canvas function. 😞 It's like it's adding another `class="canvas-container"` outside the` <canvas>` element. Any idea why? Here is the updated code:

```
export default () => {
  const fabricCanvas = useRef(new fabric.Canvas());
  const canvasEl = useRef();
  const canvasWidth = 1200;
  const canvasHeight = 1200;

  useEffect(
    () => {
      fabricCanvas.current.initialize(canvasEl.current, {
        width: canvasWidth / 2,
        height: canvasHeight / 2,
        backgroundColor: "white",
        preserveObjectStacking: true,
      });
      fabric.Object.prototype.transparentCorners = false;
      fabric.Object.prototype.cornerColor = "#ff6c14";
      fabric.Object.prototype.cornerStyle = "rect";
      fabric.Object.prototype.cornerSize = "8";
      fabric.Object.prototype.borderColor = "#ff6c14";
    },
    [canvasEl.current]
  );

  useEffect(() => {
    window.addEventListener("keyup", (e) => {
      const keycode = e.keyCode;
      if (keycode === 8 || keycode === 46) {
        const canvas = fabricCanvas.current;
        const activeObjects = canvas.getActiveObjects();
        if (activeObjects[0].isEditing) return;
        canvas.remove(...activeObjects);
      }
    });
  }, []);

  const changeCanvasSize = () => {
    const canvas = fabricCanvas.current;
    canvas.setWidth(1080);
    canvas.setHeight(1920);
    canvas.calcOffset();
  };

  return (
    <div>
            <button onClick={changeCanvasSize}>
              Change Size
            </button>
          <canvas id="canvas">
            Your browser doesn't support this. You should consider upgrading.
          </canvas>
</div>
);
};
```

Here's how it looks in the browser when i click the button for resizing the canvas:

```
<div class="canvas-container" style="width: 1080px; height: 1920px; position: relative; user-select: none;">
<div class="canvas-container" style="width: 600px; height: 600px; position: relative; user-select: none;">
<canvas id="canvas" class="lower-canvas" width="1200" height="1200" style="position: absolute; width: 600px; height: 600px; left: 0px; top: 0px; touch-action: none; user-select: none;">Your browser doesn't support this. You should consider upgrading.</canvas>
<canvas class="upper-canvas " width="1200" height="1200" style="position: absolute; width: 600px; height: 600px; left: 0px; top: 0px; touch-action: none; user-select: none; cursor: default;"></canvas>
</div></div>
```

**Reply by asturur** [reply]
i may have suggested wrongly with the later initialize, is probably wrapping the canvas twice. But i m sure there is something else we can call that works

**Reply by assentorp** [reply]
Hmm fair, does that mean that useEffect is not the way forward? :-)

**Comment by asturur** [comment]
no no that's ok. i need to look better at why you have a double wrapper

**Reply by assentorp** [reply]
Cool cool, thanks for taking the time!

I also found out meanwhile that when i use the keyboard shortcuts for this code (eg. deleting using backspace):

```
useEffect(() => {
    window.addEventListener("keyup", (e) => {
      const keycode = e.keyCode;
      if (keycode === 8 || keycode === 46) {
        const canvas = fabricCanvas.current;
        const activeObjects = canvas.getActiveObjects();
        if (activeObjects[0].isEditing) return;
        canvas.remove(...activeObjects);
      }
    });
  }, []);
```

It will throw an error for "isEditing" if there is nothing selected on the canvas.

`TypeError: Cannot read property 'isEditing' of null`

**Reply by asturur** [reply]
yeah that is standard javascript issue, you have just to add another IF. nothing specific to fabricJS.

```
if (!activeObjects[0] || activeObjects[0].isEditing) return;
```

**Reply by assentorp** [reply]
Thanks @asturur - yeah i tried making an "if" check but I forgot the the activeObjects[0] 😞 . The shortcut works now... really cool! Do you know what to do with that double canvas bug?
